pkg/kube/krt/nestedjoinmerge.go

```diff
@@ -0,0 +1,362 @@
+			stop:           o.stop,
+		},
+		collections: ics,
+		regs:        make(map[collectionUID]HandlerRegistration),
+	}
+
+	j.mergejoin.collections = j
+	j.syncer = channelSyncer{
+		name:   j.collectionName,
+		synced: j.synced,
+	}
+
+	maybeRegisterCollectionForDebugging(j, o.debugger)
+
+	// Create our queue. When it syncs (that is, all items that were present when Run() was called), we mark ourselves as synced.
+	j.queue = queue.NewWithSync(func() {
+		close(j.synced)
+		j.log.Infof("%v synced (uid %v)", j.name(), j.uid())
+	}, j.collectionName)
+
+	// Subscribe to when collections are added or removed.
+	// Don't run existing because we want to ensure the first set
+	// of collections passed to us are synced before we mark
+	// ourselves as synced. Do this before returning so collections
+	// aren't added between now and when runQueue is called.
+	subscriptionFunc := func(events []Event[T]) {
+		j.queue.Push(func() error {
+			j.onSubCollectionEventHandler(events)
+			return nil
+		})
+	}
+	reg := j.collections.RegisterBatch(func(e []Event[Collection[T]]) {
+		for _, ev := range e {
+			obj := ev.Latest()
+			switch ev.Event {
+			case controllers.EventAdd:
+				// When a collection is added, subscribe to its events
+				reg := obj.RegisterBatch(subscriptionFunc, true)
+				j.mu.Lock()
+				j.regs[obj.(internalCollection[T]).uid()] = reg
+				j.mu.Unlock()
+			case controllers.EventUpdate:
+				j.handleCollectionUpdate(ev)
+			case controllers.EventDelete:
+				j.handleCollectionDelete(ev)
+			}
+		}
+	}, false)
+	initialCollections := j.collections.List()
+	// Finally, async wait for the primary to be synced. Once it has, we know it has enqueued the initial state.
+	// After this, we can run our queue.
+	// The queue will process the initial state and mark ourselves as synced (from the NewWithSync callback)
+	go j.runQueue(initialCollections, subscriptionFunc, reg)
+
+	return j
+}
+
+func (j *nestedjoinmerge[T]) runQueue(initialCollections []Collection[T], subscriptionFunc func([]Event[T]), reg HandlerRegistration) {
+	// Wait for the container of collections to be synced before we start processing events.
+	if !j.collections.WaitUntilSynced(j.stop) {
+		return
+	}
+	j.mu.Lock()
+
+	// Now that we've subscribed, process the current set of collections.
+	for _, c := range initialCollections {
+		// Save these registrations so we can unsubscribe later if the collection is deleted.
+		// Ensure each sub-collection is synced before we're marked as synced.
+		j.regs[c.(internalCollection[T]).uid()] = c.RegisterBatch(subscriptionFunc, true)
+	}
+
+	regs := append([]HandlerRegistration{reg}, maps.Values(j.regs)...)
+	j.mu.Unlock()
+
+	syncers := slices.Map(regs, func(r HandlerRegistration) cache.InformerSynced {
+		return r.HasSynced
+	})
+
+	if !kube.WaitForCacheSync(j.collectionName, j.stop, syncers...) {
+		return
+	}
+	j.queue.Run(j.stop)
+}
+
+func (j *nestedjoinmerge[T]) handleCollectionUpdate(e Event[Collection[T]]) {
```

Comment:
If the collection is updated, then it could be a completely new collection with new event handlers, so we would have to remove handlers form the old collection and add them to the new collection.